src/main/java/uk/gov/hmcts/ccd/domain/service/search/elasticsearch/CrossCaseTypeSearchRequest.java

```diff
@@ -45,12 +56,75 @@ public JsonNode getSearchRequestJsonNode() {
+        return multiCaseTypeSearch;
     }
 
     private void validateJsonSearchRequest() {
         if (!searchRequestJsonNode.has(QUERY)) {
             throw new BadSearchRequest("missing required field 'query'");
         }
     }
+
+    public List<String> getSourceFilterAliasFields() {
+        return sourceFilterAliasFields;
+    }
+
+    public static class Builder {
+
+        private final List<String> caseTypeIds = new ArrayList<>();
+        private JsonNode searchRequestJsonNode;
+        private boolean multiCaseTypeSearch;
+        private final List<String> sourceFilterAliasFields = new ArrayList<>();
+
+        public Builder withCaseTypes(List<String> caseTypeIds) {
+            if (caseTypeIds != null) {
+                this.caseTypeIds.addAll(caseTypeIds);
+            }
+            return this;
+        }
+
+        public Builder withSearchRequest(JsonNode searchRequestJsonNode) {
+            this.searchRequestJsonNode = searchRequestJsonNode;
+            return this;
+        }
+
+        public Builder withMultiCaseTypeSearch(boolean multiCaseTypeSearch) {
+            this.multiCaseTypeSearch = multiCaseTypeSearch;
+            if (multiCaseTypeSearch && searchRequestJsonNode.has(SOURCE)) {
+                setSourceFilterAliasFields(searchRequestJsonNode.get(SOURCE));
+            }
+            return this;
+        }
+
+        public Builder withSourceFilterAliasFields(List<String> sourceFilterAliasFields) {
+            if (sourceFilterAliasFields != null) {
+                this.sourceFilterAliasFields.addAll(sourceFilterAliasFields);
+            }
+            return this;
+        }
+
+        private void setSourceFilterAliasFields(JsonNode multiCaseTypeSearchSourceNode) {
+            if (multiCaseTypeSearchSourceNode != null && multiCaseTypeSearchSourceNode.isArray()) {
+                // Alias fields are expected in source filter for multi-case types searches. If no aliases are defined in source filter, only meta data will
+                // be returned with no data fields. If aliases are defined in the source filter, alias fields with data and metadata will be returned.
+                sourceFilterAliasFields.addAll(StreamSupport.stream(multiCaseTypeSearchSourceNode.spliterator(), false)
+                                                   .map(JsonNode::asText)
+                                                   .filter(nodeText -> nodeText.startsWith(CROSS_CASE_TYPE_SEARCH_ALIAS_FIELD_PREFIX))
+                                                   .map(nodeText -> nodeText.replaceFirst(CROSS_CASE_TYPE_SEARCH_ALIAS_FIELD_PREFIX, ""))
+                                                   .collect(Collectors.toList()));
+                // remove source filter defined for multi case type search
+                removeSourceFilter();
+            }
+        }
+
+        private void removeSourceFilter() {
+            ((ObjectNode) this.searchRequestJsonNode).remove(SOURCE);
```

Comment:
if the source contains a mix of alias and non alias fields, we are removing also the non alias fields from the query right?
We are in the end not returning them to the client because we filter them in the postfiltering, but we still get them from ES. 
Can't we leave the non alias fields in the _source so ES does that part of the filtering for us?

Reply:
if we leave non alias fields in the _source, ES will return only these fields and wont return original fields for the aliases so expecting ES to return all fields